inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/annotation/DynamicAnnotatorWorkflowActionBarItemGroup.java

```diff
@@ -87,50 +107,99 @@ public String getStateClass()
-                            "ORDER BY rand()";
-            for (SourceDocument doc: entityManager.createQuery(query,SourceDocument.class)
-                .setParameter("project", project).getResultList()) {
-                //Check if it even exists or is state NEW
-                if (documentService.listAnnotatableDocuments(project,user).get(doc) == null ||
-                    documentService.listAnnotatableDocuments(project,user).get(doc).
-                        getState().equals(NEW)) {
-                    getAnnotationPage().getModelObject().setDocument(doc, documentService.
-                        listSourceDocuments(project));
-                    //This document had the state NEW, load it
-                    getAnnotationPage().actionLoadDocument(_target);
-                    _target.add(page);
-                    return;
-                }
+                    ANNOTATION_IN_PROGRESS_TO_ANNOTATION_FINISHED);
+
+            List<AnnotationDocument> inProgressDocuments = workloadManagementService
+                    .getAnnotationDocumentListForUserWithState(project, user, IN_PROGRESS);
+
+            // Assign a new document with actionLoadDocument
+
+            // First, check if there are other documents which have been in the state INPROGRESS
+            // Load the first one found
+            if (!inProgressDocuments.isEmpty()) {
+                getAnnotationPage().getModelObject().setDocument(
+                        inProgressDocuments.get(0).getDocument(),
+                        documentService.listSourceDocuments(project));
+                getAnnotationPage().actionLoadDocument(aTarget);
+                return;
             }
 
-            //Check if no new document has been selected, return to homepage
-            if (getAnnotationPage().getModelObject().getDocument().equals(document)) {
-                getAnnotationPage().setResponsePage(getAnnotationPage().getApplication().
-                    getHomePage());
-                getSession().info("There are no more documents to annotate available for you. Please contact your project supervisor.");
+            // No annotation documents in the state INPROGRESS, now select a new one
+            // depending on the workload strategy selected
+            WorkloadManager currentWorkload = workloadManagementService
+                    .loadOrCreateWorkloadManagerConfiguration(project);
+
+            // Get all documents for which the state is NEW, or which have not been created yet.
+            List<SourceDocument> sourceDocuments = workloadManagementService
+                    .getAnnotationDocumentListForUser(project, user);
+
+            // Switch for all workflow types which are available. If a new one is created.
+            // simply add a new case here.
+            switch (dynamicWorkloadExtension.readTraits(currentWorkload).getWorkflowType()) {
+
+            // Go through all documents in a random order
+            case (RANDOMIZED_WORKFLOW):
```

Comment:
Instead of doing a switch here, there should be a method in the `WorkflowExtension` interface like `SourceDocument nextDocument(List<SourceDocument> candidates)` and that method should be used to fetch the next document. The specific implementations like the `RandomizedWorkflowExtension` implement this method accordingly. The menu bar code should not have to know anything about the selection process except that it calls the `nextDocument` method.